src/System.Diagnostics.PerformanceCounter/src/System/Diagnostics/PerformanceCounterLib.cs

```diff
@@ -299,19 +301,17 @@ internal static void CloseAllLibraries()
         {
             if (s_libraryTable != null)
             {
-                foreach (PerformanceCounterLib library in s_libraryTable.Values)
-                    library.Close();
-
-                s_libraryTable = null;
-            }
-        }
+                //race with GetPerformanceCounterLib
```

Comment:
Does this locking actually make a material difference in race conditions with GetPerformanceCounterLib?  GetPerformanceCounterLib doesn't take the lock on InternalSyncObj if s_libraryTable is already initialized.  And even if it did, it would only take the lock long enough to initialize the table, but it would be handing out a reference to a PerformanceCounterLib that we're closing here, so code could be using a PerformanceCounterLib concurrently with it being Close'd.  Is that ok?  Or maybe the locking we're adding here is purely about protecting the mutation of s_libraryTable, and we're not concerned with the other issues?

Reply:
>Or maybe the locking we're adding here is purely about protecting the mutation of s_libraryTable, and we're not concerned with the other issues?

Yes @stephentoub testing with parallel Register/Unregister(CloseAllLibraries) and GetCounterXXX i see a lot of `new PerformanceCounterLib(machineName, lcidString); ` on GetPerformanceCounterLib for `s_libraryTable = null;` this lock is only to protect mutation s_libraryTable and avoid a lot of benign instances of PerformanceCounterLib on parallel scenario

>but it would be handing out a reference to a PerformanceCounterLib that we're closing here

doesn't seem to be a problem because the "registry tables" will be reloaded, if counter is no more present a not found exception will raise.
Again is a borderline scenario(parallel tests with a lot of register/unregister in parallel).